### PR TITLE
[netatmo] Make OAuth2 token refresh RFC compliant

### PR DESCRIPTION
**Pull request by clinique**
Cherry picking PR #14548 in branch 3.4.x

### CODE REVIEW

**Comment by jlaur** [comment]
@clinique - with 3.4.2 and a JAR built from this branch, when I go to `http://<your openHAB address>:8080/netatmo/connect/<_CLIENT_ID_>` and click "Authorize Thing", I'm taken to a Netatmo page. When I here click "Yes, I accept", I get this result:

```
HTTP ERROR 500 java.lang.NoSuchMethodError: 'java.lang.String java.lang.String.formatted(java.lang.Object[])'
URI:	/netatmo/connect/<client_id>
STATUS:	500
MESSAGE:	java.lang.NoSuchMethodError: 'java.lang.String java.lang.String.formatted(java.lang.Object[])'
SERVLET:	org.ops4j.pax.web.service.spi.model.ServletModel-64
CAUSED BY:	java.lang.NoSuchMethodError: 'java.lang.String java.lang.String.formatted(java.lang.Object[])'
Caused by:
java.lang.NoSuchMethodError: 'java.lang.String java.lang.String.formatted(java.lang.Object[])'
	at org.openhab.binding.netatmo.internal.api.AuthenticationApi.requestToken(AuthenticationApi.java:99)
	at org.openhab.binding.netatmo.internal.api.AuthenticationApi.authorize(AuthenticationApi.java:74)
	at org.openhab.binding.netatmo.internal.handler.ApiBridgeHandler.openConnection(ApiBridgeHandler.java:167)
	at org.openhab.binding.netatmo.internal.servlet.GrantServlet.handleRedirect(GrantServlet.java:124)
	at org.openhab.binding.netatmo.internal.servlet.GrantServlet.doGet(GrantServlet.java:82)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:687)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:799)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:550)
	at org.ops4j.pax.web.service.jetty.internal.HttpServiceServletHandler.doHandle(HttpServiceServletHandler.java:74)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:600)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:235)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1624)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1440)
	at org.ops4j.pax.web.service.jetty.internal.HttpServiceContext.doHandle(HttpServiceContext.java:294)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:501)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1594)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1355)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.ops4j.pax.web.service.jetty.internal.JettyServerHandlerCollection.handle(JettyServerHandlerCollection.java:90)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	at org.eclipse.jetty.server.Server.handle(Server.java:516)
	at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:487)
	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:732)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:479)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:277)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
	at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:338)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:315)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:173)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:131)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:409)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:883)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1034)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

Do you know what might be the issue?

**Comment by jlaur** [comment]
@clinique - oh, I just realized I compiled the JAR with Java 17. `String.formatted` was introduced in Java 15, so can't be used with 3.4.

**Comment by lolodomo** [comment]
I believe @clinique introduced code that is not compatible with Java 11 (formatted).

File: bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/api/AuthenticationApi.java [hunk 1]
```diff
@@ -51,69 +52,74 @@ public class AuthenticationApi extends RestManager {
...
             } catch (NetatmoException e) {
                 logger.warn("Unable to refresh access token : {}", e.getMessage());
             }
-        }, Math.round(response.getExpiresIn() * 0.8), TimeUnit.SECONDS));
-        tokenResponse = Optional.of(response);
-        return response.getRefreshToken();
+        }, Math.round(response.getExpiresIn() * 0.9), TimeUnit.SECONDS));
+
+        grantedScope = response.getScope();
+        authorization = "Bearer %s".formatted(response.getAccessToken());
```
Comment (jlaur) [inline bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/api/AuthenticationApi.java:99]:
```suggestion
        authorization = "Bearer " + response.getAccessToken();
```

**Review by jlaur — APPROVED** [review]
Thanks!

**Comment by jlaur** [comment]
Regarding test, I've been running this version since the PR was created. We probably can't test more until Netatmo actually makes the changes.

**Comment by jlaur** [comment]
@kaikreuzer - because of other (conflicting) Netatmo changes in the main branch, we backported this fix through this dedicated PR. I don't know how you create the release notes for patch releases, but this one should be mentioned. I don't know if milestone should be set to anything on this PR.

**Comment by kaikreuzer** [comment]
Thanks @jlaur. I am actually pulling all issues with 4.0 milestone and "patch" label for the release notes. I've therefore just added the "patch" label to #14548 as well, so that we can see on that PR directly that it has been backported (in this case through a dedicated PR instead of being cherry-picked).

**Pull request merged**